node/actors/bft/src/leader/replica_commit.rs

```diff
@@ -118,37 +118,32 @@ impl StateMachine {
 
         // ----------- All checks finished. Now we process the message. --------------
 
-        // TODO: we have a bug here since we don't check whether replicas commit
-        // to the same proposal.
-
         // We add the message to the incrementally-constructed QC.
-        self.commit_qcs
+        let commit_qc = self
+            .commit_qcs
             .entry(message.view.number)
-            .or_insert_with(|| CommitQC::new(message.clone(), self.config.genesis()))
-            .add(&signed_message, self.config.genesis());
+            .or_default()
+            .entry(message.clone())
+            .or_insert_with(|| CommitQC::new(message.clone(), self.config.genesis()));
+        commit_qc.add(&signed_message, self.config.genesis());
 
         // We store the message in our cache.
-        let cache_entry = self
-            .commit_message_cache
+        self.commit_message_cache
```

Comment:
do we still need it? It duplicates data with commit_qcs

Reply:
It works just as before 🤷‍♂️. It aggregates by view and validator while `commit_qcs` aggregates by view and (now) by message.

I guess we can obtain the validator from the CommitQC signers, but that's not optimal and also could have been done with previous `commit_qc` version.

Reply:
I mean commit_message_cache unnecessarily contains the messages. It would be enough if it just contained the sets of validators - or even better, we could have a validator -> view map which will just indicate the highest view of messages that we have received from the given validator. And yes, I know it was suboptimal before, but we can fix it.

Reply:
Ok, solved this checking if the author signature is in the qc.